src/System.Net.Http/src/System/Net/Http/Managed/AuthenticateAndRedirectHandler.cs

```diff
@@ -0,0 +1,231 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Net.Http.Headers;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http
+{
+    internal sealed partial class AuthenticateAndRedirectHandler : HttpMessageHandler
+    {
+        private readonly HttpMessageHandler _innerHandler;
+        private readonly bool _preAuthenticate;
+        private readonly ICredentials _credentials;
+        private readonly bool _allowRedirect;
+        private readonly int _maxAutomaticRedirections;
+        private AuthenticationHelper.DigestResponse _digestResponse;
+
+        public AuthenticateAndRedirectHandler(bool preAuthenticate, ICredentials credentials, bool allowRedirect, int maxAutomaticRedirections, HttpMessageHandler innerHandler)
+        {
+            Debug.Assert(innerHandler != null);
+
+            _preAuthenticate = preAuthenticate;
+            _credentials = credentials;
+            _allowRedirect = allowRedirect;
+
+            if (allowRedirect && maxAutomaticRedirections < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(maxAutomaticRedirections));
+            }
+
+            _maxAutomaticRedirections = maxAutomaticRedirections;
+            _innerHandler = innerHandler;
+        }
+
+        protected internal override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            HttpResponseMessage response;
+            uint redirectCount = 0;
+            while (true)
+            {
+                // Just as with WinHttpHandler and CurlHandler, for security reasons, we drop the server credential if it is
+                // anything other than a CredentialCache on redirection. We allow credentials in a CredentialCache since they
+                // are specifically tied to URIs.
+                ICredentials currentCredential = redirectCount > 0 ? _credentials as CredentialCache : _credentials;
+
+                if (currentCredential != null && _preAuthenticate)
+                {
+                    // Try using previous digest response WWWAuthenticate header
+                    if (_digestResponse != null)
+                    {
+                        await AuthenticationHelper.TrySetDigestAuthToken(request, currentCredential, _digestResponse, HttpKnownHeaderNames.Authorization).ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        AuthenticationHelper.TrySetBasicAuthToken(request, currentCredential);
+                    }
+                }
+
+                response = await _innerHandler.SendAsync(request, cancellationToken).ConfigureAwait(false);
+
+                if (currentCredential != null && !_preAuthenticate && response.StatusCode == HttpStatusCode.Unauthorized)
```

Comment:
If it used the wrong digest challenge for preauthenticate then it won't retry here with the new challenge.

Reply:
This logic existed before, when preauthenticate is set, the handler sends auth data with prior information. If it is not authenticated then the error is thrown to user. Also again, I think you are confusing about the cache value here. The cache value here just has the information about the authscheme.parameter value from previous challenge from server. For every request the auth response from client is recalculated, the only thing that is constant or cached here is credentials, realm, algorithm, qop information from server on prior challenge.